File: 2022/day15/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

function parseSensors(text) {
  const sensors = [];
  const pattern = /Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)/;

  for (const line of text.split('\n')) {
    const match = line.match(pattern);
    const sx = parseInt(match[1]);
    const sy = parseInt(match[2]);
    const bx = parseInt(match[3]);
    const by = parseInt(match[4]);
    const dist = Math.abs(sx - bx) + Math.abs(sy - by);
    sensors.push({ sx, sy, bx, by, dist });
  }

  return sensors;
}

function mergeRanges(ranges) {
  if (ranges.length === 0) return [];

  ranges.sort((a, b) => a[0] - b[0]);
  const merged = [ranges[0]];

  for (let i = 1; i < ranges.length; i++) {
    const [start, end] = ranges[i];
    const last = merged[merged.length - 1];
    if (start <= last[1] + 1) {
      last[1] = Math.max(last[1], end);
    } else {
      merged.push([start, end]);
    }
  }

  return merged;
}

function getCoverageAtRow(sensors, row) {
  const ranges = [];

  for (const { sx, sy, dist } of sensors) {
    const rowDist = Math.abs(sy - row);
    if (rowDist > dist) continue;

    const xSpread = dist - rowDist;
    ranges.push([sx - xSpread, sx + xSpread]);
  }

  return mergeRanges(ranges);
}

function part1() {
  const sensors = parseSensors(input);
  const targetRow = 2000000;

  const ranges = getCoverageAtRow(sensors, targetRow);

  // Count total coverage
  let total = 0;
  for (const [start, end] of ranges) {
    total += end - start + 1;
  }

  // Subtract beacons on this row
  const beaconsOnRow = new Set();
  for (const { bx, by } of sensors) {
    if (by === targetRow) {
      beaconsOnRow.add(bx);
    }
  }

  return total - beaconsOnRow.size;
}

function part2() {
  const sensors = parseSensors(input);
  const maxCoord = 4000000;

  for (let row = 0; row <= maxCoord; row++) {
    const ranges = getCoverageAtRow(sensors, row);

    // Clip ranges to search area
    const clipped = [];
    for (const [start, end] of ranges) {
      if (end < 0 || start > maxCoord) continue;
      clipped.push([Math.max(0, start), Math.min(maxCoord, end)]);
    }

    const merged = mergeRanges(clipped);

    // Check if full row is covered
    if (merged.length === 1 && merged[0][0] === 0 && merged[0][1] === maxCoord) {
      continue;
    }

    // Found a gap
    let x;
    if (merged.length > 1) {
      x = merged[0][1] + 1;
    } else if (merged[0][0] > 0) {
      x = 0;
    } else {
      x = merged[0][1] + 1;
    }

    return BigInt(x) * 4000000n + BigInt(row);
  }

  return null;
}

console.log('Part 1:', part1());
console.log('Part 2:', part2().toString());

File: 2022/day15/rust/solution.rs
use std::fs;
use std::path::Path;

#[derive(Debug)]
struct Sensor {
    sx: i64,
    sy: i64,
    bx: i64,
    by: i64,
    dist: i64,
}

fn parse_sensors(text: &str) -> Vec<Sensor> {
    let mut sensors = Vec::new();

    for line in text.lines() {
        if line.is_empty() {
            continue;
        }

        // Extract all numbers (including negative) from the line
        let numbers: Vec<i64> = line
            .split(|c: char| !c.is_ascii_digit() && c != '-')
            .filter(|s| !s.is_empty())
            .filter_map(|s| s.parse().ok())
            .collect();

        if numbers.len() >= 4 {
            let sx = numbers[0];
            let sy = numbers[1];
            let bx = numbers[2];
            let by = numbers[3];
            let dist = (sx - bx).abs() + (sy - by).abs();
            sensors.push(Sensor { sx, sy, bx, by, dist });
        }
    }

    sensors
}

fn get_coverage_at_row(sensors: &[Sensor], row: i64) -> Vec<(i64, i64)> {
    let mut ranges = Vec::new();

    for sensor in sensors {
        let row_dist = (sensor.sy - row).abs();
        if row_dist > sensor.dist {
            continue;
        }

        let x_spread = sensor.dist - row_dist;
        ranges.push((sensor.sx - x_spread, sensor.sx + x_spread));
    }

    merge_ranges(&mut ranges)
}

fn merge_ranges(ranges: &mut Vec<(i64, i64)>) -> Vec<(i64, i64)> {
    if ranges.is_empty() {
        return Vec::new();
    }

    ranges.sort();
    let mut merged = vec![ranges[0]];

    for &(start, end) in &ranges[1..] {
        let last = merged.last_mut().unwrap();
        if start <= last.1 + 1 {
            last.1 = last.1.max(end);
        } else {
            merged.push((start, end));
        }
    }

    merged
}

fn part1(sensors: &[Sensor]) -> i64 {
    let target_row = 2000000;
    let ranges = get_coverage_at_row(sensors, target_row);

    // Count total coverage
    let total: i64 = ranges.iter().map(|(start, end)| end - start + 1).sum();

    // Subtract beacons on this row
    let mut beacons_on_row: Vec<i64> = sensors
        .iter()
        .filter(|s| s.by == target_row)
        .map(|s| s.bx)
        .collect();
    beacons_on_row.sort();
    beacons_on_row.dedup();

    total - beacons_on_row.len() as i64
}

fn part2(sensors: &[Sensor]) -> i64 {
    let max_coord = 4000000;

    for row in 0..=max_coord {
        let ranges = get_coverage_at_row(sensors, row);

        // Clip ranges to search area
        let mut clipped: Vec<(i64, i64)> = ranges
            .iter()
            .filter(|&&(start, end)| end >= 0 && start <= max_coord)
            .map(|&(start, end)| (start.max(0), end.min(max_coord)))
            .collect();

        let clipped = merge_ranges(&mut clipped);

        // Check if full row is covered
        if clipped.len() == 1 && clipped[0].0 == 0 && clipped[0].1 == max_coord {
            continue;
        }

        // Found a gap - the beacon is in the gap
        let x = if clipped.len() > 1 {
            clipped[0].1 + 1
        } else if clipped[0].0 > 0 {
            0
        } else {
            clipped[0].1 + 1
        };

        return x * 4000000 + row;
    }

    0
}

fn main() {
    let exe_path = std::env::current_exe().unwrap();
    let exe_dir = exe_path.parent().unwrap();
    let input_path = exe_dir.join("../input.txt");

    // Try relative path first, then fall back to finding input.txt
    let text = if input_path.exists() {
        fs::read_to_string(&input_path).unwrap()
    } else {
        // Try from current directory
        let alt_path = Path::new("../input.txt");
        if alt_path.exists() {
            fs::read_to_string(alt_path).unwrap()
        } else {
            // Try looking in expected location
            fs::read_to_string("/Users/wrb/fun/code/agenticadvent/2022/day15/input.txt").unwrap()
        }
    };

    let sensors = parse_sensors(&text);

    println!("Part 1: {}", part1(&sensors));
    println!("Part 2: {}", part2(&sensors));
}
